lib/ret/storage.ex

```diff
@@ -337,6 +337,111 @@ defmodule Ret.Storage do
   defp check_promotion_token(actual_token, token) when actual_token != token,
     do: {:error, :invalid_key}
 
+  # Due to a bug, OwnedFile that are not referred from anywhere can be remained
+  # as active in the database then the associated files are not removed.
+  # This function finds and deletes such files.
+  # The function might be removed when the root issue will be resolved.
+  # TODO: Where (In what file) should this function be defined?
+  # Question: How should this function be invoked? Cron? Manually?
+  def cleanup_nonaccessed_files do
+    Logger.info("Stored Files: Attempting to clean up non-referred files.")
+
+    import Ecto.Query, only: [from: 2]
+
+    # TODO: Optimize, if possible
+    # TODO: Error handling if needed
+    # TODO: Atomic operation
+    # Question: Lock the database to avoid conflict operations?
+
+    # 1. Mark all OwnedFiles as inactive
+
+    # Question: Should this be in owned_file.ex?
+    Repo.update_all(OwnedFile, set: [state: :inactive])
+
+    # 2. Mark the OwnedFile as active that referred by RoomObjects
+
+    # file_id is encoded in RoomObject.gltf_node.extensions.HUBS_components.media.src as like
+    # https://hubs.local:4000/files/71301c23-8fb3-4388-8eec-83b3627f9ad0.glb?token=xxx
+    from(object in Ret.RoomObject, select: object.gltf_node)
+      |> Repo.all()
+      |> Enum.map(&Jason.decode!/1)
+      |> Enum.each(fn (gltf_node) ->
+           # Question: Is it guaranteed that this property exists?
+           src = gltf_node["extensions"]["HUBS_components"]["media"]["src"]
+           # Question: Is it possible that src doesn't follow the format written above?
+           file_id = Regex.run(~r/files\/([^.]*)\./, src)
```

Comment:
If the `RoomObject` does _not_ point to an uploaded file, then the src will be a url to a public resource (`https://www.example.com/some_image.png`).

Reply:
Thanks, updated the code to take into such urls account